pydantic/main.py

```diff
@@ -627,7 +627,10 @@ def copy(
         cls = self.__class__
         m = cls.__new__(cls)
         object_setattr(m, '__dict__', v)
-        object_setattr(m, '__fields_set__', self.__fields_set__.copy())
+        # new `__fields_set__` can have unset optional fields with a set value in `update` kwarg
+        object_setattr(
+            m, '__fields_set__', {k for k in self.__fields__ if k in self.__fields_set__ | set(update or {})}
+        )
```

Comment:
```suggestion
        if update:
            fields_set = self.__fields_set__ | update.keys()
        else:
            fields_set = set(self.__fields_set__)
        object_setattr(m, '__fields_set__', fields_set)
```
You could also add ` & self.__fields__.keys()` to match your currently logic if you think it's required, but I don't think it is.

Reply:
Absolutely!